# Goodreads Reading List Recommender Project (Data Cleaning)

In [1]:
# Import basic data libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Import spacy for text preprocessing 
import spacy

# Import helper functions from functions.py
import functions

## Importing and Inspecting

In [2]:
# Load the reviews data into pd.DataFrame objects using pandas and set the date columns to date dtypes 
reviews = pd.read_csv(
    '../Data/goodreads_reviews.csv',
    parse_dates=['started_at', 'read_at', 'date_added']
)

# Load the works data into pd.DataFrame
books = pd.read_csv(
    '../Data/goodreads_works.csv',
    dtype={'isbn': str, 'isbn13': str}
)

In [3]:
# Make sure the reviews table loaded in correctly by checking the head
reviews.head(3)

review_id                           user_id  \
0  fa7a00c01296e3b2b2e857d79c51ea77  3693bb4f1062b659a354848cf11ca313   
1  de0f7c8d15e247443e51969becf2878e  3693bb4f1062b659a354848cf11ca313   
2  e79b49504ef58b2defcdc8b79e2ec3fb  3693bb4f1062b659a354848cf11ca313   

   work_id started_at read_at date_added  rating  \
0  6128277        NaN     NaN 2013-12-21     5.0   
1  3270810        NaN     NaN 2013-12-21     5.0   
2  4768235        NaN     NaN 2013-12-19     5.0   

                                         review_text  n_votes  n_comments  
0  Fire is half-human and half monster. Monsters ...        0           0  
1  Katsa is a graceling - blessed with an ability...        0           0  
2  This is a fun, light-hearted read. Tammy Jo is...        0           0

In [4]:
# Make sure the books table loaded in correctly by checking the head
books.head(3)

work_id        isbn         isbn13         original_title  \
0   2919130  1416534601  9781416534600              Nocturnes   
1  52087333         NaN            NaN              Draw Play   
2   1649583  1416505520  9781416505525  Citizen of the Galaxy   

               author  original_publication_year  num_pages  \
0       John Connolly                     2004.0        NaN   
1           Tia Lewis                     2016.0        NaN   
2  Robert A. Heinlein                     1957.0        NaN   

                                         description  \
0                                                NaN   
1  Jake:\nI can't believe my coach assigned me a ...   
2  In a distant galaxy, the atrocity of slavery w...   

                                              genres  \
0  fiction, fantasy, paranormal, mystery, thrille...   
1                                   romance, fiction   
2  fiction, young-adult, fantasy, paranormal, chi...   

                                           image_url  reviews_count  \
0  https://s.gr-assets.com/assets/nophoto/book/11...           8820   
1  https://s.gr-assets.com/assets/nophoto/book/11...           2482   
2  https://s.gr-assets.com/assets/nophoto/book/11...          16506   

   text_reviews_count  5_star_ratings  4_star_ratings  3_star_ratings  \
0                 338            1118            1601            1029   
1                 204             204             353             274   
2                 447            3539            4351            2863   

   2_star_ratings  1_star_ratings  ratings_count  avg_rating similar_books  
0             190              58           3996         3.9           NaN  
1              77              29            937         3.7           NaN  
2             444              53          11250         4.0           NaN

## Cleaning and Wrangling

For each table, investigate and clean the following: 
- Data types will need to be converted
- Bit-sizes can be decreased
- Possible outliers may exist in numeric columns
- Missing data may exist
- Duplicates may exists
- Inconsistent text and typos may exist in categorical columns 

### Reviews Table

In [5]:
# Inspect the data types of the reviews table
reviews.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1143887 entries, 0 to 1143886
Data columns (total 10 columns):
 #   Column       Non-Null Count    Dtype         
---  ------       --------------    -----         
 0   review_id    1143887 non-null  object        
 1   user_id      1143887 non-null  object        
 2   work_id      1143887 non-null  int64         
 3   started_at   796392 non-null   object        
 4   read_at      1031765 non-null  object        
 5   date_added   1143887 non-null  datetime64[ns]
 6   rating       1106569 non-null  float64       
 7   review_text  1143887 non-null  object        
 8   n_votes      1143887 non-null  int64         
 9   n_comments   1143887 non-null  int64         
dtypes: datetime64[ns](1), float64(1), int64(3), object(5)
memory usage: 1.6 GB


The `started_at` and `read_at` columns should be datetime64 dtypes. All other data types look correct.

In [6]:
# Convert the started_at and read_at columns to datetime64 dtypes and turn all values that can not be parsed into NaT values
reviews['started_at'] = pd.to_datetime(reviews['started_at'], errors='coerce')
reviews['read_at'] = pd.to_datetime(reviews['read_at'], errors='coerce')

In [7]:
# Check to see if the changes occured correctly
reviews.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1143887 entries, 0 to 1143886
Data columns (total 10 columns):
 #   Column       Non-Null Count    Dtype         
---  ------       --------------    -----         
 0   review_id    1143887 non-null  object        
 1   user_id      1143887 non-null  object        
 2   work_id      1143887 non-null  int64         
 3   started_at   796391 non-null   datetime64[ns]
 4   read_at      1031763 non-null  datetime64[ns]
 5   date_added   1143887 non-null  datetime64[ns]
 6   rating       1106569 non-null  float64       
 7   review_text  1143887 non-null  object        
 8   n_votes      1143887 non-null  int64         
 9   n_comments   1143887 non-null  int64         
dtypes: datetime64[ns](3), float64(1), int64(3), object(3)
memory usage: 1.4 GB


We lost one value of the `started_at` column and two values of the `read_at` column using errors='coerce' in pd.to_datetime.

In [8]:
# Check the summary statistics for the numeric columns to see if bit-sizes can be reduced and identify possible outliers
reviews.describe().round()

work_id                     started_at  \
count   1143887.0                         796391   
mean   17830071.0  2014-10-25 17:59:29.161881344   
min         104.0            1916-01-01 00:00:00   
25%     3443248.0            2013-05-27 00:00:00   
50%    15956441.0            2015-01-24 00:00:00   
75%    25032500.0            2016-05-31 00:00:00   
max    57717521.0            2017-12-29 00:00:00   
std    14795716.0                            NaN   

                             read_at                     date_added  \
count                        1031763                        1143887   
mean   2014-05-27 04:23:58.279140096  2014-01-25 03:31:19.395080192   
min              1915-05-26 00:00:00            2006-08-29 00:00:00   
25%              2013-01-25 00:00:00            2012-07-31 00:00:00   
50%              2014-10-14 00:00:00            2014-03-29 00:00:00   
75%              2016-04-01 00:00:00            2015-10-22 00:00:00   
max              2017-12-31 00:00:00            2017-10-29 00:00:00   
std                              NaN                            NaN   

          rating    n_votes  n_comments  
count  1106569.0  1143887.0   1143887.0  
mean         4.0        3.0         1.0  
min          1.0       -3.0        -1.0  
25%          3.0        0.0         0.0  
50%          4.0        0.0         0.0  
75%          5.0        1.0         0.0  
max          5.0     3222.0       833.0  
std          1.0       18.0         5.0

* The `rating`, `n_votes`, and `n_comments` columns take values between -3 and 3222. I will reduce the bit-sizes for those columns to the minimum bit-size to reduce memory usage and optimise efficiency.
* Looking at the Goodreads website, you can not dislike a review. Therefore, the `n_votes` and `n_comments` columns likely have outliers since there are negative values. I will turn the negative values in these columns to 0.

In [9]:
# Turn the negative values in n_votes and n_comments to 0
reviews['n_votes'] = np.where(reviews['n_votes'] < 0, 0, reviews['n_votes'])
reviews['n_comments'] = np.where(reviews['n_comments'] < 0, 0, reviews['n_comments'])

In [10]:
# Check to see if the changes worked correctly
reviews.describe().round().loc['min',['n_votes', 'n_comments']]

n_votes       0.0
n_comments    0.0
Name: min, dtype: object

In [11]:
# Reduce the bit-sizes of the above stated columns
reviews = reviews.astype({
    'rating': 'float32',
    'n_votes': 'int16',
    'n_comments': 'int16'
})

In [12]:
# Check to see if the changes worked correctly
reviews.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1143887 entries, 0 to 1143886
Data columns (total 10 columns):
 #   Column       Non-Null Count    Dtype         
---  ------       --------------    -----         
 0   review_id    1143887 non-null  object        
 1   user_id      1143887 non-null  object        
 2   work_id      1143887 non-null  int64         
 3   started_at   796391 non-null   datetime64[ns]
 4   read_at      1031763 non-null  datetime64[ns]
 5   date_added   1143887 non-null  datetime64[ns]
 6   rating       1106569 non-null  float32       
 7   review_text  1143887 non-null  object        
 8   n_votes      1143887 non-null  int16         
 9   n_comments   1143887 non-null  int16         
dtypes: datetime64[ns](3), float32(1), int16(2), int64(1), object(3)
memory usage: 1.4 GB


In [13]:
# Check the value_counts of ratings to see if there are any anomolies
reviews['rating'].value_counts()

rating
4.0    396722
5.0    336246
3.0    242109
2.0     94030
1.0     37462
Name: count, dtype: int64

No anomolies in the `rating` column.

In [14]:
# Check the percentage of missing data for each column
(reviews.isna().sum() / reviews.shape[0]).round(2)

review_id      0.00
user_id        0.00
work_id        0.00
started_at     0.30
read_at        0.10
date_added     0.00
rating         0.03
review_text    0.00
n_votes        0.00
n_comments     0.00
dtype: float64

**NOTE TO SELF:**
* We can fill in the rating missing values with the sentiment of the review_text.
* If we care about how long it takes to read, is there any point in having a read_at value if we do not know the started_at value

**ALSO:**
* There are no text categorical columns

In [15]:
# Check for duplicates in the whole dataset
reviews.duplicated().sum()

np.int64(0)

In [16]:
# Check for duplicates in the review_id column
reviews.duplicated(subset=['review_id']).sum()

np.int64(0)

In [17]:
# Check if the same user reviewed the same book more than once
reviews.duplicated(subset=['user_id', 'work_id']).sum()

np.int64(2323)

In [18]:
# Inspect the rows where a user reviewed the same book more than once
reviews[reviews.duplicated(subset=['user_id', 'work_id'], keep=False)].sort_values(by='user_id').head(2)

review_id                           user_id  \
529654  5173fae9e68df8cea6200ed7c7b1d875  00608f34a76da8ba3ed37b25f263adf4   
669062  2b05746c58293271e3cd76b9d66892df  00608f34a76da8ba3ed37b25f263adf4   

         work_id started_at    read_at date_added  rating  \
529654  12803237 2013-07-25 2013-09-14 2013-07-22     4.0   
669062  12803237 2013-09-05 2013-09-14 2013-09-05     4.0   

                                              review_text  n_votes  n_comments  
529654  This was my first book from Laura Buzo and sho...        0           0  
669062  This was my first book from Laura Buzo and sho...        1           0

**NOTE TO SELF:**

* Export to SQL to further investigate the duplicates

### Books Table

In [19]:
# Inspect the data types of the books table
books.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13525 entries, 0 to 13524
Data columns (total 20 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   work_id                    13525 non-null  int64  
 1   isbn                       11474 non-null  object 
 2   isbn13                     11864 non-null  object 
 3   original_title             13525 non-null  object 
 4   author                     13525 non-null  object 
 5   original_publication_year  13507 non-null  float64
 6   num_pages                  12795 non-null  float64
 7   description                13356 non-null  object 
 8   genres                     13525 non-null  object 
 9   image_url                  13525 non-null  object 
 10  reviews_count              13525 non-null  int64  
 11  text_reviews_count         13525 non-null  int64  
 12  5_star_ratings             13525 non-null  int64  
 13  4_star_ratings             13525 non-null  int

All data types look correct.

In [20]:
# Check the summary statistics for the numeric columns to see if bit-sizes can be reduced and identify possible outliers
books.describe().round()

work_id  original_publication_year  num_pages  reviews_count  \
count     13525.0                    13507.0    12795.0        13525.0   
mean   18436259.0                     2003.0      335.0        61441.0   
std    16258712.0                       59.0      138.0       203063.0   
min         104.0                     -500.0        1.0          107.0   
25%     2953328.0                     2006.0      260.0         6804.0   
50%    15992791.0                     2011.0      334.0        16588.0   
75%    25878877.0                     2014.0      391.0        44017.0   
max    57717521.0                     2021.0     2201.0      6057595.0   

       text_reviews_count  5_star_ratings  4_star_ratings  3_star_ratings  \
count             13525.0         13525.0         13525.0         13525.0   
mean               2037.0         14398.0         12172.0          6903.0   
std                5429.0         70546.0         45515.0         25036.0   
min                  18.0             5.0             4.0             4.0   
25%                 286.0           759.0           947.0           596.0   
50%                 664.0          2343.0          2650.0          1576.0   
75%                1709.0          7925.0          7854.0          4463.0   
max              156575.0       3131920.0       1519190.0        808753.0   

       2_star_ratings  1_star_ratings  ratings_count  avg_rating  
count         13525.0         13525.0        13525.0     13525.0  
mean           1914.0           833.0        36220.0         4.0  
std            8474.0          5781.0       148374.0         0.0  
min               0.0             0.0           36.0         2.0  
25%             146.0            44.0         2673.0         4.0  
50%             383.0           123.0         7351.0         4.0  
75%            1100.0           369.0        22025.0         4.0  
max          444888.0        463808.0      5066596.0         5.0

* `original_publication_year` could have an outlier since it has a value of -500.
* `num_pages` may have outliers since there is a book with a single page.

In [21]:
# Investigate the book with a release year of -500
books[books['original_publication_year'] == -500]

work_id        isbn         isbn13                  original_title  \
10917  3200649  1590302257  9781590302255  Sun Zi Bing Fa  [Sunzi bingfa]   

        author  original_publication_year  num_pages  \
10917  Sun Tzu                     -500.0      273.0   

                                             description  \
10917  Conflict is an inevitable part of life, accord...   

                                                  genres  \
10917  non-fiction, history, historical fiction, biog...   

                                               image_url  reviews_count  \
10917  https://images.gr-assets.com/books/1453417993m...         469134   

       text_reviews_count  5_star_ratings  4_star_ratings  3_star_ratings  \
10917                6631           73919           73612           50069   

       2_star_ratings  1_star_ratings  ratings_count  avg_rating similar_books  
10917           12038            3074         212712         4.0           NaN

After research, this book was released 500 BC, meaning -500 in the `original_publication_year` column is not an anomoly.

In [22]:
# Investigate the books with the most reviews
books.sort_values('num_pages', ascending=True).head(3)

work_id isbn isbn13      original_title       author  \
2509    6793408  NaN    NaN  A Study in Emerald  Neil Gaiman   
12299  43401121  NaN    NaN         Annie's Day    Andy Weir   
12122  22025856  NaN    NaN        Other People  Neil Gaiman   

       original_publication_year  num_pages  \
2509                      2003.0        1.0   
12299                     2011.0        3.0   
12122                     2001.0        3.0   

                                             description  \
2509   Alluding to both the Sherlock Holmes canon and...   
12299  I didn't want to wake up at 6am, especially on...   
12122  Free online fiction.\n"Time is fluid here," sa...   

                                                  genres  \
2509   fantasy, paranormal, mystery, thriller, crime,...   
12299              fiction, romance, fantasy, paranormal   
12122  fantasy, paranormal, fiction, mystery, thrille...   

                                               image_url  reviews_count  \
2509   https://s.gr-assets.com/assets/nophoto/book/11...          13290   
12299  https://images.gr-assets.com/books/1481823872m...           1882   
12122  https://images.gr-assets.com/books/1458233173m...            746   

       text_reviews_count  5_star_ratings  4_star_ratings  3_star_ratings  \
2509                  337            2570            2025             819   
12299                 455             254             412             283   
12122                  51             167             125              70   

       2_star_ratings  1_star_ratings  ratings_count  avg_rating  \
2509              156              54           5624         4.2   
12299              80              41           1070         3.7   
12122               6               5            373         4.2   

                                           similar_books  
2509                                             4593256  
12299  44734637, 22025856, 15268747, 21883429, 185992...  
12122                       43401121, 21883429, 16703904

**NOTE TO SELF**:

* Goodreads says that 'A Study in Emerald' is one page. However, Google says 88 pages

In [23]:
# Check the percentage of missing data for each column
(books.isna().sum() / books.shape[0]).round(2)

work_id                      0.00
isbn                         0.15
isbn13                       0.12
original_title               0.00
author                       0.00
original_publication_year    0.00
num_pages                    0.05
description                  0.01
genres                       0.00
image_url                    0.00
reviews_count                0.00
text_reviews_count           0.00
5_star_ratings               0.00
4_star_ratings               0.00
3_star_ratings               0.00
2_star_ratings               0.00
1_star_ratings               0.00
ratings_count                0.00
avg_rating                   0.00
similar_books                0.19
dtype: float64

**NOTE TO SELF:**
* Figure out what to do with missing values later

**ALSO:**
* There are no text categorical columns

In [24]:
# Check for duplicates in the whole dataset
books.duplicated().sum()

np.int64(0)

In [25]:
# Check for duplicates book titles
books.duplicated(subset=['original_title', 'author']).sum()

np.int64(12)

In [26]:
books[books.duplicated(subset=['original_title', 'author'], keep=False)].sort_values('author')

work_id        isbn         isbn13  \
12283  16736232         NaN            NaN   
12289  15456915         NaN            NaN   
2985    1460966  0689867042  9780689867040   
4326     588218  0689868235  9780689868238   
1797   11163588  1606841750  9781606841754   
9683   21699195  1455857408  9781455857401   
931      385372  0061059927  9780061059926   
3345     285261  1847386823  9781847386823   
11042   7071165  0553269216  9780553269215   
1423    1133797  0553213180  9780553213188   
2704   13344769  0375714839  9780375714832   
3390    3303888  037571457X  9780375714573   
2923     594714  0007118899  9780007118892   
3954    2098474  0007162499  9780007162499   
3372      15983  0553564943  9780553564945   
10744   2960376  0586217835  9780586217832   
258    17329122  142154105X  9781421541051   
2419   15949685  1421541041  9781421541044   
315    17329123  1421541068  9781421541068   
12171  13009740         NaN            NaN   
12621  40295208         NaN            NaN   
658    50388100  1563127881  9781563127885   
3986    2902834  0675093244  9780675093248   

                                        original_title            author  \
12283                                           Pulled    Amy Lichtenhan   
12289                                           Pulled    Amy Lichtenhan   
2985                       Tithe: A Modern Faerie Tale       Holly Black   
4326                       Tithe: A Modern Faerie Tale       Holly Black   
1797                                          Monsters      Ilsa J. Bick   
9683                                          Monsters      Ilsa J. Bick   
931                                       Dark Reunion        L.J. Smith   
3345                                      Dark Reunion        L.J. Smith   
11042                                   Rainbow Valley   L.M. Montgomery   
1423                                    Rainbow Valley   L.M. Montgomery   
2704                                        Persepolis   Marjane Satrapi   
3390                                        Persepolis   Marjane Satrapi   
2923   Tatiana and Alexander (The Bronze Horseman, #2)   Paullina Simons   
3954   Tatiana and Alexander (The Bronze Horseman, #2)   Paullina Simons   
3372                                          Magician  Raymond E. Feist   
10744                                         Magician  Raymond E. Feist   
258                                    Reimi no Arcana          Rei Toma   
2419                                   Reimi no Arcana          Rei Toma   
315                                    Reimi no Arcana          Rei Toma   
12171                                         Evenfall   Santino Hassell   
12621                                         Evenfall   Santino Hassell   
658                                   A Rose for Emily  William Faulkner   
3986                                  A Rose for Emily  William Faulkner   

       original_publication_year  num_pages  \
12283                     2011.0      242.0   
12289                     2011.0      388.0   
2985                      2002.0      332.0   
4326                      2002.0      314.0   
1797                      2011.0      465.0   
9683                      2012.0        NaN   
931                       1991.0      287.0   
3345                      1991.0      732.0   
11042                     1919.0      256.0   
1423                      1917.0      243.0   
2704                      2003.0      341.0   
3390                      2000.0      153.0   
2923                      2003.0      559.0   
3954                      2006.0      742.0   
3372                      1982.0      485.0   
10744                     1982.0      681.0   
258                       2011.0      192.0   
2419                      2009.0      192.0   
315                       2011.0      189.0   
12171                     2007.0     1427.0   
12621                     2014.0      344.0   
658                       1930.0       36.0 

These are most likely updated versions of the same book. For now I will simply leave this.

I will need to clean the `genres` and `similar_books` columns by putting one value per column

In [27]:
# Turn each value in the genres column into a list
books['genres'] = books['genres'].str.split(', ')

# Create dummy columns for each genre and save it as a pd.DataFrame so it can be used as a lookup table
genres = (pd.get_dummies(
    books.explode('genres')[['work_id', 'genres']],
    columns=['genres'],
    prefix='',
    prefix_sep=''
               ).groupby('work_id')
                .sum()
                .reset_index()
)

# Inspect the genres table
genres.head()

work_id  biography  children  comics  crime  fantasy  fiction  graphic  \
0      104          1         0       0      0        0        1        0   
1      114          1         0       0      0        1        1        0   
2      115          1         1       0      0        0        1        0   
3      423          0         0       0      0        1        1        0   
4      434          1         0       0      1        0        1        0   

   historical fiction  history  mystery  non-fiction  paranormal  poetry  \
0                   1        1        0            1           0       0   
1                   1        1        0            0           1       0   
2                   1        1        0            0           0       0   
3                   0        0        0            0           1       0   
4                   1        1        1            1           0       0   

   romance  thriller  young-adult  
0        1         0            0  
1        1         0            0  
2        0         0            1  
3        0         0            0  
4        0         1            0

In [28]:
# Check to see if the dummy variables worked correctly
genres.max()

work_id               57717521
biography                    2
children                     2
comics                       2
crime                        2
fantasy                      2
fiction                      2
graphic                      2
historical fiction           2
history                      2
mystery                      2
non-fiction                  1
paranormal                   2
poetry                       1
romance                      2
thriller                     2
young-adult                  2
dtype: int64

In [29]:
# Set max column width to None (no limit) so that I can see the whole text
pd.set_option('display.max_colwidth', None)

# Inspect a row that has 2 for dummy variable
books[books['work_id'] == int(genres[genres['biography'] == 2]['work_id'].iloc[0])]['genres']

3826    [fantasy, paranormal, history, historical fiction, biography, mystery, thriller, crime, fiction, comics, graphic, young-adult, fantasy, paranormal, history, historical fiction, biography, mystery, thriller, crime, fiction, comics, graphic]
Name: genres, dtype: object

There are duplicates in some of the genres lists. Therefore, replace 2 with 1 for every dummy variable.

In [30]:
# Check if any work_id equal 2 so its not replaced with 1
any(books['work_id'] == 2)

False

In [31]:
# Replace all 2 with 1 in the 
genres = genres.replace(2, 1)

# Check if the change occured correctly
genres.max()

work_id               57717521
biography                    1
children                     1
comics                       1
crime                        1
fantasy                      1
fiction                      1
graphic                      1
historical fiction           1
history                      1
mystery                      1
non-fiction                  1
paranormal                   1
poetry                       1
romance                      1
thriller                     1
young-adult                  1
dtype: int64

------------------------------------------------------------------------------------------------------------------------------

In [32]:
# Turn each value in the genres column into a list
books['similar_books'] = books['similar_books'].str.split(', ')

# Create dummy columns for each genre and save it as a pd.DataFrame so it can be used as a lookup table
similar_books_temp = (pd.get_dummies(
    books.explode('similar_books')[['work_id', 'similar_books']],
    columns=['similar_books'],
    prefix='',
    prefix_sep=''
               ).groupby('work_id')
                .sum()
                .reset_index()
)

# Check the shape of similar books
similar_books_temp.shape

(13525, 12116)

In [33]:
# Make the index work_id
similar_books_temp = similar_books_temp.set_index('work_id')

In [34]:
# See how many missing books there are
similar_books_temp.shape[0] - similar_books_temp.shape[1]

1410

In [35]:
# Create arrays of the index and columns that are of the same dtype and have the same value type
index = np.array(similar_books_temp.index, dtype='object').astype(str)
columns = np.array(similar_books_temp.columns, dtype='object')

# See how many work_id in columns that are in the index
matching = [i for i in columns if i in index]
len(matching)

8560

There are columns (work_id) that are not in the dataset. There are 8560 that actually exist in the dataset.

In [36]:
# Remove columns that do not exist in the data
similar_books_temp = similar_books_temp[matching]

# Chack to see if the changes occures correctly
similar_books_temp.shape

(13525, 8560)

In [37]:
# Find the missing books (columns) in the similar_books_temp table
existing_columns = np.array(similar_books_temp.columns, dtype='object')
missing_columns = [str(i) for i in index if i not in existing_columns]

In [38]:
# Join the missing columns onto the similar_books_temp df with all zeros 
missing_similar_books_data = pd.DataFrame(0, index=similar_books_temp.index, columns=missing_columns)
similar_books = pd.concat([similar_books_temp, missing_similar_books_data], axis=1)

In [39]:
# Sort the columns to be symmetric to the index (ascending order by work_id)
similar_books = similar_books[np.array(sorted(similar_books.columns.astype(int).tolist())).astype(str)]

# Remove the name of the index
similar_books.index.name = None

# Inspect the similar_books table
similar_books.head()

104  114  115  423  434  505  696  797  817  860  ...  56825332  \
104    0    0    0    0    0    0    0    0    0    0  ...         0   
114    0    0    0    0    0    0    0    0    0    0  ...         0   
115    0    0    0    0    0    0    0    0    0    0  ...         0   
423    0    0    0    0    0    0    0    0    0    0  ...         0   
434    0    0    0    0    0    0    0    0    0    0  ...         0   

     56847346  56947505  57104739  57400201  57407230  57414860  57438069  \
104         0         0         0         0         0         0         0   
114         0         0         0         0         0         0         0   
115         0         0         0         0         0         0         0   
423         0         0         0         0         0         0         0   
434         0         0         0         0         0         0         0   

     57646853  57717521  
104         0         0  
114         0         0  
115         0         0  
423         0         0  
434         0         0  

[5 rows x 13525 columns]

In [40]:
# Drop the genres and similar_books column from the original books table
books = books.drop(['genres', 'similar_books'], axis=1)

### Feature Engineering

In [41]:
# Create a column with the value of how many days it took the user to read the book
reviews['days_to_finish'] = reviews['read_at'] - reviews['started_at']

In [42]:
# Get the average amount of days users took to read each book
mean_days_to_finish = (reviews.groupby('work_id')['days_to_finish'].mean()
                              .dt.days
                              .reset_index()
                              .rename({'days_to_finish': 'avg_days_to_finish'}, axis=1))

# Create a column in the books table with the average amount of days users took to read each book
books = books.merge(mean_days_to_finish, on='work_id')

### Text Preprocessing

In [43]:
# Use this cell to toggle between setting display.max_colwidth to None and resetting it

# Set max column width to None (no limit) so that I can see the whole text
pd.set_option('display.max_colwidth', None)

# Reset the max column width so the pd.DataFrame is more compact
# pd.reset_option('display.max_colwidth')

In [44]:
# Remove ' \n' and '\n' from the text in the review_text column (inspecting the values, replacing with '')
reviews['review_text'] = reviews['review_text'].str.replace(r'\s*\n', '', regex=True)

In [45]:
# Inspect some review_text values with spoilers in it
reviews[reviews['review_text'].apply(lambda x: '(view spoiler)' in x)]['review_text'].head(2)

230    Wow! This book was pretty intense. There was quite a bit going on in this novel compared to the others before it. (view spoiler)[For one, we learn a lot more about Rafe and his unique abilities, not to mention the other members of the team. We also learn that Rafe is in love with her and doesn't want to be. Their relationship will always be complicated and just might get more so, as time moves on. For two, we see Violet get stronger powers, she is actually able to see when a person dies. At least that is one big scene in there. We also see Violet fighting for control over her relationship. She has feelings for Rafe but her boyfriend Jay is the one who has taken up her heart for so long. We see her fight with herself, wondering if she should give up Jay, her job and many other complicated situations. For three, we deal with Violet being kidnapped by a serial killer. How she tries to survive during the whole thing and after. When she kills him, she picks up an 'echo' of her own an

In [46]:
# Create a new column with the spoiler part of the review_text
reviews['spoiler_text'] = (reviews['review_text'].str.extract(r'\[(.*?)\]', expand=False)
                                                 .str.replace('(hide spoiler)', '', regex=False)
                                                 .str.strip())

In [47]:
# Create a new column with review text without the spoiler parts (sf = spoiler free)
reviews['review_text_sf'] = (reviews['review_text'].str.replace('(view spoiler)', '')
                                                   .str.replace(r'\[(.*?)\]', '', regex=True))

In [48]:
# Load the spacy english model
nlp = spacy.load('en_core_web_sm')

# Helper functions for text preprocessing (data cleaning) section
def lower_replace(series):
    output = series.str.lower()
    output = output.str.replace(r'[^\w\s]', '', regex=True)
    return output

def clean_and_normalise(series):
    # Step 1: lowercase & remove punctuation (fast, vectorized)
    cleaned = lower_replace(series)
    
    # Step 2: process texts in batches with nlp.pipe()
    processed_texts = nlp.pipe(cleaned, batch_size=1000)
    
    # Step 3: lemma and remove stopwords for each doc
    output = []
    for doc in processed_texts:
        tokens = [token.lemma_ for token in doc if not token.is_stop]
        output.append(' '.join(tokens))
    
    return pd.Series(output)

In [49]:
# Clean and normalise the review_text_sf column with the helper function
# reviews['cleaned_review_text'] = functions.clean_and_normalise(reviews['review_text_sf'])

In [50]:
# Clean and normalise the spoiler_text column with the helper function
# reviews['cleaned_spoiler_text'] = functions.clean_and_normalise(reviews['spoiler_text'])

In [51]:
# Clean and normalise the description column from the books table with the helper function
books['description_clean'] = functions.clean_and_normalise(books['description'])

100%|███████████████████████████████████████████████████████████████████████████████████| 13525/13525 [39:29<00:00,  5.71it/s]


------------------------------------------------------------------------------------------------------------------------------

## Exporting

In [52]:
# Export the books, reviews, genres, and similar_books tables as pickle files
pd.to_pickle(books, '../Data/books.pkl')
pd.to_pickle(reviews, '../Data/reviews.pkl')
pd.to_pickle(genres, '../Data/genres.pkl')
pd.to_pickle(similar_books, '../Data/similar_books.pkl')